from prep_inputs import *
from utils import *
#from functions import *
import random
from pathlib import Path

# top_dir = Path(config['top_dir'])
# working_dir = top_dir / config['cohort_name'] / f"{config['sample_name']}_{config['genome_version']}" 

workdir: working_dir # <----- working directory is specified by prep_inputs.py

rule all:
    input:
        # expand("mutect2_sc_pass1/m2_sc_vcfs/{sample_name}_{cell_barcode}_somatic_m2.vcf.gz", cell_barcode = bars, sample_name = sample_name),
        # expand("mutect2_sc_pass1/m2_sc_vcfs_filter_added/{sample_name}_{cell_barcode}_somatic_m2_filter_added.vcf.gz", cell_barcode = bars, sample_name = sample_name),
        # expand("bcf_pass1/combined_vcf/{sample_name}-combined_VCF_filter_v3.prev_filtered.vcf.gz", sample_name = sample_name),
        expand("mutect2_sc_f_pass2/m2_sc_vcfs/{sample_name}_{cell_barcode}_somatic_m2.vcf.gz", cell_barcode = bars, sample_name = sample_name),

include: "rules/2-single-cell_mutect2_call.smk"
include: "rules/3-filter_and_combine_sc_calls.smk"
include: "rules/4-single-cell_mutect2_force_call.smk"
